### PR TITLE
Close the connections buffered for the listeners

### PR DESCRIPTION
**Pull request by soheilhy**
This commit closes the connections that has been buffered
on the connection channel of the cmux listeners, when the
root listener is closed. With this change it is guaranteed
that the connections are either closed or handed of to the
child listeners (returned via Accept()).

There are a couple of changes to the tests to cover corner
cases and the new behavior.

### CODE REVIEW

File: cmux.go [hunk 1]
```diff
@@ -97,7 +97,11 @@ func (m *cMux) Serve() error {
 		wg.Wait()
 
 		for _, sl := range m.sls {
+			// Drain the connections enqueued for the listener.
```
Comment (tamird) [inline cmux.go]: nit: put this below the `close()`

File: cmux_test.go [hunk 2]
```diff
@@ -374,9 +374,14 @@ func TestClose(t *testing.T) {
...
 
-	// Second connection goes through.
+	// Second connection either goes through or it is closed.
 	if _, err := anyl.Accept(); err != nil {
-		t.Fatal(err)
+		if err != ErrListenerClosed {
+			t.Fatal(err)
+		}
+		if _, err := c2.Read([]byte{}); err != io.ErrClosedPipe {
+			t.Fatal("connection is not closed and is leaked: %v", err)
```
Comment (tamird) [inline cmux_test.go]: s/Fatal/Fatalf/

**Comment by tamird** [comment]
LGTM

File: cmux.go (hunk 1 above)
Comment (soheilhy) [inline cmux.go]: Thanks. Fixed. :-)

**Comment by soheilhy** [comment]
Thanks @tamird!

File: cmux.go [hunk 3]
```diff
@@ -98,6 +98,10 @@ func (m *cMux) Serve() error {
 
 		for _, sl := range m.sls {
 			close(sl.l.connc)
+			// Drain the connections enqueued for the listener.
+			for c := range sl.l.connc {
+				c.Close()
```
Comment (tamird) [inline cmux.go]: heh, need `_ = c.Close()`

**Pull request merged**